File: distance-matrix/python/async_matrix.py
"""Compute a large distance matrix with the asynchronous Distance API and save it as CSV."""

from __future__ import annotations

import argparse
import csv
import os
import sys
import time
from pathlib import Path
from typing import Any

import requests

API_URL = "https://api.woosmap.com/distance/matrix/async/"
FINAL_STATUSES = {"completed", "timeout", "error"}
OUTPUT_COLUMNS = ["origin_index", "destination_index", "status", "distance_m", "duration_s"]


def retry_delay(response: requests.Response, attempt: int) -> float:
    # Woosmap sends ratelimit-reset, in seconds; Retry-After only comes from proxies
    for header in ("ratelimit-reset", "Retry-After"):
        try:
            return max(0.0, float(response.headers[header]))
        except (KeyError, ValueError):
            continue
    return float(2**attempt)


def read_points(path: Path) -> list[str]:
    points: list[str] = []
    for line in path.read_text(encoding="utf-8-sig").splitlines():
        cells = [cell.strip() for cell in line.replace(";", ",").split(",")]
        if len(cells) < 2 or not cells[0] or not cells[1]:
            continue
        try:
            float(cells[0]), float(cells[1])
        except ValueError:
            continue
        points.append(f"{cells[0]},{cells[1]}")
    if not points:
        raise ValueError(f"no coordinates found in {path}")
    return points


class AsyncMatrix:
    def __init__(self, private_key: str, session: requests.Session | None = None) -> None:
        self.private_key = private_key
        self.session = session or requests.Session()

    def _call(self, method: str, url: str, **kwargs: Any) -> requests.Response:
        for attempt in range(3):
            response = self.session.request(
                method, url, params={"private_key": self.private_key}, timeout=60, **kwargs
            )
            if response.status_code != 429 or attempt == 2:
                break
            time.sleep(retry_delay(response, attempt))
        if response.status_code >= 400:
            raise RuntimeError(f"{method} {url} failed ({response.status_code}): {response.text}")
        return response

    def submit(self, origins: list[str], destinations: list[str], **options: str) -> str:
        body = {"origins": "|".join(origins), "destinations": "|".join(destinations), **options}
        return self._call("POST", API_URL, json=body).json()["matrix_id"]

    def status(self, matrix_id: str) -> str:
        return self._call("GET", f"{API_URL}{matrix_id}/status").json()["status"]

    def result(self, matrix_id: str) -> dict[str, Any]:
        return self._call("GET", f"{API_URL}{matrix_id}").json()

    def wait(self, matrix_id: str, interval: float, timeout: float) -> str:
        deadline = time.monotonic() + timeout
        while True:
            status = self.status(matrix_id)
            print(f"{matrix_id}: {status}", file=sys.stderr)
            if status in FINAL_STATUSES:
                return status
            if time.monotonic() >= deadline:
                raise TimeoutError(f"matrix {matrix_id} still {status} after {timeout}s")
            time.sleep(interval)


def flatten(result: dict[str, Any]) -> list[dict[str, Any]]:
    # The async result is not the synchronous rows/elements shape: travelTimes and distances
    # are flat row-major arrays, errorCodes is present only when some pairs failed.
    matrix = result.get("matrix") or {}
    destinations = matrix.get("numDestinations") or 0
    times, distances = matrix.get("travelTimes") or [], matrix.get("distances") or []
    errors = matrix.get("errorCodes") or []
    rows: list[dict[str, Any]] = []
    for index in range(matrix.get("numOrigins", 0) * destinations):
        error = errors[index] if index < len(errors) else 0
        rows.append(
            {
                "origin_index": index // destinations,
                "destination_index": index % destinations,
                "status": "OK" if not error else f"ERROR_{error}",
                "distance_m": distances[index] if index < len(distances) and not error else "",
                "duration_s": times[index] if index < len(times) and not error else "",
            }
        )
    return rows


def write_csv(path: Path, rows: list[dict[str, Any]]) -> None:
    with path.open("w", newline="", encoding="utf-8") as handle:
        writer = csv.DictWriter(handle, fieldnames=OUTPUT_COLUMNS, lineterminator="\n")
        writer.writeheader()
        writer.writerows(rows)


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "output", type=Path, help="CSV to write, one row per origin/destination pair"
    )
    parser.add_argument("--origins", type=Path, help="CSV with one lat,lng per line")
    parser.add_argument("--destinations", type=Path, help="CSV with one lat,lng per line")
    parser.add_argument("--matrix-id", help="resume polling a job submitted earlier instead")
    parser.add_argument(
        "--mode", default="driving", help="driving (default), walking, cycling or truck"
    )
    parser.add_argument("--method", choices=("time", "distance"), default="time")
    parser.add_argument("--elements", default="duration_distance")
    parser.add_argument(
        "--poll-interval", type=float, default=5.0, help="seconds between status checks"
    )
    parser.add_argument("--timeout", type=float, default=1800.0, help="seconds before giving up")
    return parser


def private_key_from_env() -> str:
    key = os.environ.get("WOOSMAP_PRIVATE_KEY")
    if not key:
        raise SystemExit("set WOOSMAP_PRIVATE_KEY in the environment")
    return key


def submit_from_files(api: AsyncMatrix, args: argparse.Namespace) -> str:
    if not (args.origins and args.destinations):
        raise SystemExit("pass --origins and --destinations, or --matrix-id to resume a job")
    origins, destinations = read_points(args.origins), read_points(args.destinations)
    matrix_id = api.submit(
        origins, destinations, mode=args.mode, method=args.method, elements=args.elements
    )
    # the job outlives this process: keep the id to resume with --matrix-id if polling is cut
    print(f"submitted {len(origins)}x{len(destinations)} matrix {matrix_id}", file=sys.stderr)
    return matrix_id


def main(argv: list[str] | None = None) -> int:
    args = build_parser().parse_args(argv)
    api = AsyncMatrix(private_key_from_env())
    matrix_id = args.matrix_id or submit_from_files(api, args)
    status = api.wait(matrix_id, args.poll_interval, args.timeout)
    if status != "completed":
        print(f"matrix {matrix_id} ended with status {status}", file=sys.stderr)
        return 1
    rows = flatten(api.result(matrix_id))
    write_csv(args.output, rows)
    print(f"wrote {len(rows)} elements to {args.output}", file=sys.stderr)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: distance-matrix/node/async-matrix.mjs
// Compute a large distance matrix with the asynchronous Distance API and save it as CSV.
import { readFile, writeFile } from "node:fs/promises";
import { pathToFileURL } from "node:url";
import { parseArgs } from "node:util";

export const API_URL = "https://api.woosmap.com/distance/matrix/async/";
const FINAL_STATUSES = new Set(["completed", "timeout", "error"]);
const OUTPUT_COLUMNS = ["origin_index", "destination_index", "status", "distance_m", "duration_s"];

const sleep = (seconds) => new Promise((resolve) => setTimeout(resolve, seconds * 1000));

export function parsePoints(text) {
  const points = [];
  for (const line of text.replace(/^﻿/, "").split(/\r?\n/)) {
    const [lat, lng] = line.replace(/;/g, ",").split(",").map((c) => c.trim());
    if (lat && lng && !Number.isNaN(Number(lat)) && !Number.isNaN(Number(lng))) points.push(`${lat},${lng}`);
  }
  if (!points.length) throw new Error("no coordinates found");
  return points;
}

// IETF RateLimit header: comma-separated "policy";r=<remaining>;t=<reset-seconds> entries
function parseRateLimit(value) {
  return value
    .split(",")
    .filter((policy) => policy.trim())
    .map((policy) => {
      const result = {};
      for (const match of policy.matchAll(/\b([rt])=(\d+)/g)) result[match[1]] = Number(match[2]);
      return result;
    });
}

// a 429 is bound by whichever policy hit zero, not necessarily the first one in the header;
// ratelimit-reset is a compat header pending removal, Retry-After only ever comes from a proxy
function retryDelay(response, attempt) {
  const policies = parseRateLimit(response.headers.get("RateLimit") ?? "");
  const exhausted = policies.filter((p) => p.r === 0 && p.t !== undefined).map((p) => p.t);
  if (exhausted.length) return Math.max(...exhausted);
  for (const header of ["ratelimit-reset", "retry-after"]) {
    const raw = response.headers.get(header);
    const value = raw?.trim() ? Number(raw) : Number.NaN;
    if (Number.isFinite(value) && value >= 0) return value;
  }
  return 2 ** attempt;
}

export class AsyncMatrix {
  constructor(privateKey, fetchImpl = fetch, sleepImpl = sleep, now = () => Date.now() / 1000) {
    this.privateKey = privateKey;
    this.fetch = fetchImpl;
    this.sleep = sleepImpl;
    this.now = now;
  }

  async call(method, url, body) {
    const target = `${url}?private_key=${encodeURIComponent(this.privateKey)}`;
    let response;
    for (let attempt = 0; attempt < 3; attempt += 1) {
      response = await this.fetch(target, {
        method,
        headers: body ? { "Content-Type": "application/json" } : {},
        body: body ? JSON.stringify(body) : undefined,
        redirect: "follow",
      });
      if (response.status !== 429 || attempt === 2) break;
      await this.sleep(retryDelay(response, attempt));
    }
    if (!response.ok) throw new Error(`${method} ${url} failed (${response.status}): ${await response.text()}`);
    return response.json();
  }

  async submit(origins, destinations, options = {}) {
    const body = { origins: origins.join("|"), destinations: destinations.join("|"), ...options };
    return (await this.call("POST", API_URL, body)).matrix_id;
  }

  status = async (matrixId) => (await this.call("GET", `${API_URL}${matrixId}/status`)).status;
  result = (matrixId) => this.call("GET", `${API_URL}${matrixId}`);

  async wait(matrixId, interval, timeout) {
    const deadline = this.now() + timeout;
    for (;;) {
      const status = await this.status(matrixId);
      console.error(`${matrixId}: ${status}`);
      if (FINAL_STATUSES.has(status)) return status;
      if (this.now() >= deadline) throw new Error(`matrix ${matrixId} still ${status} after ${timeout}s`);
      await this.sleep(interval);
    }
  }
}

export function flatten(result) {
  // The async result is not the synchronous rows/elements shape: travelTimes and distances
  // are flat row-major arrays, errorCodes is present only when some pairs failed.
  const matrix = result.matrix ?? {};
  const destinations = matrix.numDestinations ?? 0;
  const { travelTimes = [], distances = [], errorCodes = [] } = matrix;
  const rows = [];
  for (let index = 0; index < (matrix.numOrigins ?? 0) * destinations; index += 1) {
    const error = errorCodes[index] ?? 0;
    rows.push({
      origin_index: Math.floor(index / destinations),
      destination_index: index % destinations,
      status: error ? `ERROR_${error}` : "OK",
      distance_m: error ? "" : distances[index] ?? "",
      duration_s: error ? "" : travelTimes[index] ?? "",
    });
  }
  return rows;
}

export const toCsv = (rows) =>
  [OUTPUT_COLUMNS.join(","), ...rows.map((row) => OUTPUT_COLUMNS.map((c) => row[c]).join(","))].join("\n") + "\n";

async function submitFromFiles(api, values) {
  if (!values.origins || !values.destinations) throw new Error("pass --origins and --destinations, or --matrix-id to resume a job");
  const origins = parsePoints(await readFile(values.origins, "utf8"));
  const destinations = parsePoints(await readFile(values.destinations, "utf8"));
  const matrixId = await api.submit(origins, destinations, { mode: values.mode, method: values.method, elements: values.elements });
  // the job outlives this process: keep the id to resume with --matrix-id if polling is cut
  console.error(`submitted ${origins.length}x${destinations.length} matrix ${matrixId}`);
  return matrixId;
}

export async function main(argv) {
  const { values, positionals } = parseArgs({
    args: argv,
    allowPositionals: true,
    options: {
      mode: { type: "string", default: "driving" },
      method: { type: "string", default: "time" },
      elements: { type: "string", default: "duration_distance" },
      "poll-interval": { type: "string", default: "5" },
      timeout: { type: "string", default: "1800" },
      origins: { type: "string" },
      destinations: { type: "string" },
      "matrix-id": { type: "string" },
    },
  });
  const [outputPath] = positionals;
  if (!outputPath) throw new Error("usage: node async-matrix.mjs <output.csv> --origins o.csv --destinations d.csv | --matrix-id ID");
  const privateKey = process.env.WOOSMAP_PRIVATE_KEY;
  if (!privateKey) throw new Error("set WOOSMAP_PRIVATE_KEY in the environment");
  const api = new AsyncMatrix(privateKey);
  const matrixId = values["matrix-id"] ?? (await submitFromFiles(api, values));
  const status = await api.wait(matrixId, Number(values["poll-interval"]), Number(values.timeout));
  if (status !== "completed") {
    console.error(`matrix ${matrixId} ended with status ${status}`);
    return 1;
  }
  const rows = flatten(await api.result(matrixId));
  await writeFile(outputPath, toCsv(rows));
  console.error(`wrote ${rows.length} elements to ${outputPath}`);
  return 0;
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  main(process.argv.slice(2)).then((code) => process.exit(code), (error) => {
    console.error(error.message);
    process.exit(1);
  });
}
